File: leetcode-04/solutions/python3/0130.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board:
            return

        m = len(board)
        n = len(board[0])

        for i in range(m):
            self.dfs(board, i, 0)
            self.dfs(board, i, n - 1)

        for j in range(1, n - 1):
            self.dfs(board, 0, j)
            self.dfs(board, m - 1, j)

        for i in range(m):
            for j in range(n):
                board[i][j] = 'O' if board[i][j] == '.' else 'X'

    def dfs(self, board: List[List[str]], i: int, j: int) -> None:
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or board[i][j] != 'O':
            return

        board[i][j] = '.'

        self.dfs(board, i, j + 1)
        self.dfs(board, i, j - 1)
        self.dfs(board, i + 1, j)
        self.dfs(board, i - 1, j)

File: leetcode-04/solutions/cpp/0130.cpp
class Solution {
 public:
  void solve(vector<vector<char>>& board) {
    if (board.empty()) return;

    const int m = board.size();
    const int n = board[0].size();

    for (int i = 0; i < m; i++) {
      dfs(board, i, 0);
      dfs(board, i, n - 1);
    }

    for (int j = 1; j < n - 1; j++) {
      dfs(board, 0, j);
      dfs(board, m - 1, j);
    }

    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++) board[i][j] = board[i][j] == '.' ? 'O' : 'X';
  }

 private:
  void dfs(vector<vector<char>>& board, int i, int j) {
    if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() ||
        board[i][j] != 'O')
      return;

    board[i][j] = '.';

    dfs(board, i, j + 1);
    dfs(board, i, j - 1);
    dfs(board, i + 1, j);
    dfs(board, i - 1, j);
  }
};